### PR TITLE
Packaging

### PR DESCRIPTION
**Pull request by janrito**
Added CI and packaging configuration

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -110,6 +110,13 @@ ENV/
 env.bak/
 venv.bak/
 
+# Dotenv
+.envrc
+.dotenv
+
+# vscode
+.vscode
+
```
Comment (risktoparkrageram) [inline .gitignore:119]: could we add `.idea/` too?

**Review by risktoparkrageram — APPROVED** [review]

File: .gitignore (hunk 1 above)
Comment (janrito) [reply .gitignore:119]: 👍🏽 758290b

File: hyperframe/tabulate.py [hunk 2]
```diff
@@ -46,19 +46,19 @@ def tabulate_from_hyperpattern_tags(tags_with_structure: list[Tag]) -> pd.DataFr
...
+    258  cefccdb2-b558-5623-9c17-72b4be7bd4c1  /film/2001-a-space-odyssey           2001: A Space Odyssey  ...                    Directed by                Stanley Kubrick            USA, United Kingdom
+    259  1b8c3bcb-842f-5ba2-90ee-39b6b40b7c21  /film/in-the-mood-for-love            In the Mood for Love  ...                    Directed by                   Wong Kar Wai              Hong Kong, France
+    260  65edaa59-6546-57b5-9fa3-9409f93d1383           /film/tokyo-story                     Tokyo Story  ...                    Directed by                   Yasujirō Ozu                          Japan
+    261  f4c92833-af39-5cd6-a41e-8df5933d0dc1          /film/citizen-kane                    Citizen Kane  ...                    Directed by                   Orson Welles                            USA
+    262  489eada8-7d7e-594c-acac-5903f43f4b85               /film/vertigo                         Vertigo  ...                    Directed by               Alfred Hitchcock                            USA
+    """  # noqa: E501
```
Comment (sam-lawrence) [inline hyperframe/tabulate.py:61]: 👍

File: pyproject.toml [hunk 3]
```diff
@@ -1,24 +1,38 @@
...
 readme = "README.md"
+repository = "https://github.com/sam-lawrence/HyperFrame"
+classifiers = [
+    "Operating System :: OS Independent",
+    "Programming Language :: Python :: 3",
+    "Programming Language :: Python :: 3.9",
+    "Programming Language :: Python :: 3.10",
+    "Programming Language :: Python :: 3.11",
+    "Programming Language :: Python :: 3 :: Only",
+]
```
Comment (sam-lawrence) [inline pyproject.toml:19]: what does this help with again?

File: pyproject.toml [hunk 4]
```diff
@@ -1,24 +1,38 @@
...
+packages = [{ include = "hyperframe" }, { include = "hyperframe/py.typed" }]
```
Comment (sam-lawrence) [inline pyproject.toml:20]: Is this to stop it including `.github/workflows`??  And what does `hyperframe/py.typed` do?

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    continue-on-error: ${{ matrix.experimental }}
+    name: "Build and Test on ${{ matrix.python-version }}, experimental: ${{ matrix.experimental }}"
+    strategy:
+      fail-fast: false
+      matrix:
+        python-version: ["3.9", "3.10", "3.11"]
+        experimental: [false]
+        # include:
+        #   - python-version: "3.11"
+        #     experimental: true
```
Comment (sam-lawrence) [inline .github/workflows/ci.yml:20]: reminder

**Review by sam-lawrence — APPROVED** [review]
Is CI just free for public repos?  No matter how much it's used?

**Comment by sam-lawrence** [comment]
This is great!

File: .github/workflows/release.yml [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+      - name: Install dev dependencies
+        run: python -m pip install --upgrade pip setuptools poetry
+
+      - name: Build binary wheel and a source tarball
+        run: poetry build
+
+      - name: Get version
+        run: echo "VERSION=$(poetry version --short)" >>$GITHUB_OUTPUT
+        id: version
```
Comment (sam-lawrence) [inline .github/workflows/release.yml:27]: presumably this is the version written in the pyproject.toml file?

File: .github/workflows/release.yml [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+      - name: Tag with updated version
+        id: tag_version
+        uses: mathieudutour/github-tag-action@v6.1
+        with:
+          github_token: ${{ secrets.GITHUB_TOKEN }}
+          custom_tag: ${{ steps.version.outputs.VERSION }}
```
Comment (sam-lawrence) [inline .github/workflows/release.yml:34]: what secrets are these?

File: pyproject.toml (hunk 3 above)
Comment (janrito) [reply pyproject.toml:19]: this is the standard classification system for python packages. this is just saying that this package is python 3 only, and it works with versions 3.9-3.11

File: pyproject.toml (hunk 4 above)
Comment (janrito) [reply pyproject.toml:20]: yep, it's a list of modules to include in the built distribution https://python-poetry.org/docs/pyproject/#packages

py.typed marks this package as a typed package

File: .github/workflows/ci.yml (hunk 5 above)
Comment (janrito) [reply .github/workflows/ci.yml:20]: oh, i've been leaving this in order to start testing on python 3.12 when it comes out as experimental

File: .github/workflows/release.yml (hunk 6 above)
Comment (janrito) [reply .github/workflows/release.yml:27]: yep

File: .github/workflows/release.yml (hunk 7 above)
Comment (janrito) [reply .github/workflows/release.yml:34]: it's the default github token assigned to every action, it allows the action to push a release

Comment (janrito) [reply .github/workflows/release.yml:34]: https://docs.github.com/en/actions/security-guides/automatic-token-authentication

**Comment by janrito** [comment]
> Is CI just free for public repos? No matter how much it's used?

it's free for public repos
https://github.blog/2019-08-08-github-actions-now-supports-ci-cd/
https://github.com/pricing

**Comment by janrito** [comment]
Thanks!

**Pull request merged**